R/dgp-lib.R

```diff
@@ -197,48 +144,71 @@ xy_dgp_constructor <- function(x_fun, y_fun, err_fun = NULL, add_err = TRUE,
 linear_gaussian_dgp <- function(n, p_obs = 0, p_unobs = 0, 
                                 s_obs = p_obs, s_unobs = p_unobs,
                                 betas = NULL, betas_unobs = NULL,
-                                betas_sd = 1, betas_unobs_sd = 1, 
                                 intercept = 0, err = NULL, 
                                 data_split = FALSE, train_prop = 0.5,
                                 return_values = c("X", "y", "support"),
                                 ...) {
   return_values <- match.arg(return_values, several.ok = TRUE)
-  
+
+  fun_args <- dots_to_fun_args(
+    prefix = c("X", "U", "y", "err", "betas", "betas_unobs"), ...
+  )
+  X_args_list <- fun_args$.X_args
+  U_args_list <- fun_args$.U_args
+  y_args_list <- fun_args$.y_args
+  err_args_list <- fun_args$.err_args
+  betas_args_list <- fun_args$.betas_args
+  betas_unobs_args_list <- fun_args$.betas_unobs_args
+  optional_args_list <- fun_args$.optional_args
+
   # simulate observed covariates
   if (p_obs != 0) {
-    X <- generate_X_gaussian(n = n, p = p_obs)
-    betas <- generate_coef(betas = betas, p = p_obs, s = s_obs, sd = betas_sd)
+    X <- R.utils::doCall(
+      generate_X_gaussian, n = n, p = p_obs,
+      alwaysArgs = X_args_list
+    )
+    betas <- R.utils::doCall(
+      generate_coef, .betas = betas, .p = p_obs, .s = s_obs,
+      args = c(betas_args_list, optional_args_list), .ignoreUnusedArgs = FALSE
+    )
   } else {
     X <- matrix(0, nrow = n, ncol = 1)
     betas <- 0
   }
-  
+
   # simulate unobserved covariates
   if (p_unobs != 0) {
-    U <- generate_X_gaussian(n = n, p = p_unobs)
-    betas_unobs <- generate_coef(betas = betas_unobs, 
-                                 p = p_unobs, s = s_unobs, sd = betas_unobs_sd,
-                                 betas_name = "betas_unobs")
+    U <- R.utils::doCall(
+      generate_X_gaussian, n = n, p = p_unobs,
+      alwaysArgs = U_args_list
+    )
+    betas_unobs <- R.utils::doCall(
+      generate_coef, .betas = betas_unobs, .p = p_unobs, .s = s_unobs,
+      args = c(betas_unobs_args_list, optional_args_list),
+      .ignoreUnusedArgs = FALSE
+    )
   } else {
     U <- matrix(0, nrow = n, ncol = 1)
     betas_unobs <- 0
   }
-  
+
   # simulate linear y
-  y <- generate_y_linear(
-    X = X, U = U, betas = betas, betas_unobs = betas_unobs, 
+  y <- R.utils::doCall(
+    generate_y_linear, X = X, U = U, betas = betas, betas_unobs = betas_unobs,
     intercept = intercept, err = err,
-    return_support = "support" %in% return_values, ...
+    return_support = "support" %in% return_values,
+    args = c(optional_args_list, y_args_list, err_args_list),
```

Comment:
Something worth checking here is if `y_args_list` and `err_args_list` have elements with the same names. I think this would happen if the user passed in `.y_sd` and `.err_sd` through `...`. I think in this case, both `y_args_list` and `err_args_list` would have a component named `sd`.

Reply:
OK this should be avoided now by calling `generate_errors()` inside of `linear_gaussian_dgp()`.

I think we should also make it super clear that if the _user's_ functions (`betas`, `betas_unobs`, and `err`) have common args then they _have_ to be prefixed appropriately to avoid sharing the value across their functions (though they might also want that).

Reply:
Completely agree. We should make that very clear that arguments of the same name are shared across all the user's functions. This is fine if it's intended, but they should be cautious.